From 13074109e7dc147f524f1ea566f7cc3b336d5562 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 23 Jan 2010 16:10:47 +0000
Subject: [PATCH] [CONFIGURATION-403] Fixed XMLConfiguration.load() so that an
 empty configuration that was saved and reloaded is still considered empty.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@902431 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/configuration/XMLConfiguration.java      | 10 +++++-----
 .../commons/configuration/TestXMLConfiguration.java  | 12 ++++++++++++
 xdocs/changes.xml                                    |  5 +++++
 3 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/src/java/org/apache/commons/configuration/XMLConfiguration.java b/src/java/org/apache/commons/configuration/XMLConfiguration.java
index 4542fccaca..97a2e65fe0 100644
--- a/src/java/org/apache/commons/configuration/XMLConfiguration.java
+++ b/src/java/org/apache/commons/configuration/XMLConfiguration.java
@@ -21,9 +21,9 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.Reader;
-import java.io.Writer;
-import java.io.StringWriter;
 import java.io.StringReader;
+import java.io.StringWriter;
+import java.io.Writer;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -46,9 +46,9 @@
 import javax.xml.transform.dom.DOMSource;
 import javax.xml.transform.stream.StreamResult;
 
-import org.apache.commons.configuration.tree.ConfigurationNode;
-import org.apache.commons.configuration.resolver.EntityRegistry;
 import org.apache.commons.configuration.resolver.DefaultEntityResolver;
+import org.apache.commons.configuration.resolver.EntityRegistry;
+import org.apache.commons.configuration.tree.ConfigurationNode;
 import org.apache.commons.logging.LogFactory;
 import org.w3c.dom.Attr;
 import org.w3c.dom.CDATASection;
@@ -639,7 +639,7 @@ else if (w3cNode instanceof Text)
         {
             text = text.trim();
         }
-        if (text.length() > 0 || !node.hasChildren())
+        if (text.length() > 0 || (!node.hasChildren() && node != getRoot()))
         {
             node.setValue(text);
         }
diff --git a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
index f0151c44ab..89b7a2c66d 100644
--- a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
@@ -936,6 +936,18 @@ public void testEmptyElements() throws ConfigurationException
         assertEquals("", conf.getProperty("empty2"));
     }
 
+    /**
+     * Tests the isEmpty() method for an empty configuration that was reloaded.
+     */
+    public void testEmptyReload() throws ConfigurationException
+    {
+        XMLConfiguration config = new XMLConfiguration();
+        assertTrue("Newly created configuration not empty", config.isEmpty());
+        config.save(testSaveConf);
+        config.load(testSaveConf);
+        assertTrue("Reloaded configuration not empty", config.isEmpty());
+    }
+
     /**
      * Tests whether the encoding is correctly detected by the XML parser. This
      * is done by loading an XML file with the encoding "UTF-16". If this
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 57b2649a86..4cfff37bc6 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,11 @@
 
   <body>
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-403">
+        When an empty XMLConfiguration was saved and reloaded the root element
+        was assigned an empty text value. Because of this isEmpty() returned
+        false for this configuration. This has been fixed.
+      </action>
       <action dev="oheger" type="add" issue="CONFIGURATION-399">
         Default variable interpolation now supports the env: prefix for
         referencing environment variables.
